hyperjumptech/monika
# feat(native fetch): Handle Node.js fetch errors

# Monika Pull Request (PR)

This PR resolves #1246 

## What feature/issue does this PR add

1. Handle Node.js fetch errors based on [Node.js Undici errors](https://github.com/nodejs/undici/blob/2a1721955a40ff89b4388c445998ff3353718e7a/docs/docs/api/Errors.md)

## How did you implement / how did you fix it

1. Expand low-level HTTP status code outside Axios' code range when handling Node.js fetch errors
2. Add documentation for new low-level HTTP status codes
3. Split HTTP request between using Axios and Node.js fetch
4. Split exception handling between Axios and Node.js fetch error

## How to test

1. Given monika.yml configuration
```
probes:
  - id: '2'
    name: HTML form submission
    description: simulate html form submission
    interval: 1
    requests:
      - url: https://httpbin.org/post
        timeout: 100
        method: POST
        headers:
          Accept: application/json
        body:
          username: someusername
          password: somepassword
```
5. Run `npm run start -- --native-fetch --verbose -r 1`
6. Observe log for Node.js fetch use and timeout error
![image](https://github.com/hyperjumptech/monika/assets/9563873/3100365b-a1f8-4502-a923-89620431da44)
